inference/conflict.go

```diff
@@ -198,11 +232,51 @@ func (l *conflictList) addOverconstraintConflict(nilExplanation ExplainedBool, n
 
 func (l *conflictList) diagnostics() []analysis.Diagnostic {
 	var diagnostics []analysis.Diagnostic
-	for _, c := range l.conflicts {
+
+	// group conflicts with the same nil path together for concise reporting
+	groupedConflicts := groupConflicts(l.conflicts)
+
+	// build diagnostics from grouped conflicts
+	for _, c := range groupedConflicts {
 		diagnostics = append(diagnostics, analysis.Diagnostic{
 			Pos:     c.position,
 			Message: c.String(),
 		})
 	}
 	return diagnostics
 }
+
+// groupConflicts groups conflicts with the same nil path together and update conflicts list.
+func groupConflicts(allConflicts []conflict) []conflict {
+	conflictsMap := make(map[string]int)  // key: nil path string, value: index in `allConflicts`
+	indicesToIgnore := make(map[int]bool) // indices of conflicts to be ignored from `allConflicts`, since they are grouped with other conflicts
+
+	for i, c := range allConflicts {
+		key := pathString(c.nilFlow.nilPath)
```

Comment:
Can we move the utility function `pathString` closer to this?

Reply:
Done